django/db/models/expressions.py

```diff
@@ -777,8 +777,10 @@ class When(Expression):
     def __init__(self, condition=None, then=None, **lookups):
         if lookups and condition is None:
             condition, lookups = Q(**lookups), None
-        if condition is None or not isinstance(condition, Q) or lookups:
-            raise TypeError("__init__() takes either a Q object or lookups as keyword arguments")
+        if isinstance(condition, Expression) and isinstance(condition.output_field, fields.BooleanField):
```

Comment:
Also, it probably needs to guard on `hasattr(condition, 'resolve_expression')` and `hasattr(condition, 'output_field')` if it doesn't use the `isinstance(condition, Expression)`. Or can we assume that?